### PR TITLE
Working ontoconv

### PR DESCRIPTION
**Pull request by louisponet**
I think this about covers it. I didn't yet do the tests because it makes little sense to do a 1 to 1 comparison before we're sure this is the correct output, and that OntoFlow will keep the current structure as the output. 

I've attached the resulting output files (I will clean stuff up). I think all the functionality is there but needs to be verified.

 Somehow github doesn't like yaml so it's txt files in yaml format...
[generated_workchain.txt](https://github.com/user-attachments/files/16196998/generated_workchain.txt)
[generated_pipeline_step_10.txt](https://github.com/user-attachments/files/16196999/generated_pipeline_step_10.txt)
[generated_pipeline_step_6.txt](https://github.com/user-attachments/files/16197000/generated_pipeline_step_6.txt)
[generated_pipeline_final_output.txt](https://github.com/user-attachments/files/16197001/generated_pipeline_final_output.txt)

### CODE REVIEW

**Comment by louisponet** [comment]
[pipeline_1.txt](https://github.com/user-attachments/files/16676284/pipeline_1.txt)
[pipeline_0.txt](https://github.com/user-attachments/files/16676285/pipeline_0.txt)
[workchain.txt](https://github.com/user-attachments/files/16676286/workchain.txt)

File: ontoconv/ontoflow.py [hunk 1]
```diff
@@ -9,7 +9,130 @@
 
 import yaml
 
-from ontoconv.pipelines import generate_ontoflow_pipeline
+from ontoconv.pipelines import generate_pipeline, generate_ontoflow_pipeline, load_simulation_resource
```
Comment (francescalb) [inline ontoconv/ontoflow.py]:
```suggestion
from ontoconv.pipelines import generate_ontoflow_pipeline, load_simulation_resource
```

**Review by francescalb — CHANGES_REQUESTED** [review]
The reivew probably will need several iterations, but first points are:

1. In the workchain specification for an execflow wrapper, the  'outputs' are not at the same level as the inputs, but one of the keys in the dictionary under "inputs".

2. The first sources picked up as individuals are completely missing in the first generated pipeline. I.e. these:

- downloadUrl: file://tests/testfiles/input_data/abaqus_conf.zip
  mediaType: application/vnd.dlite-parse
  type: ss3:AbaqusConfiguration
  configuration:
    options: inputfile=CementModel_Final_contact.inp;ppfile=OpenODB7.py
    metadata: http://www.sintef.no/calm/0.1/AbaqusConfiguration
    driver: abaqus_configuration
  dataresource: abaqus_config1_dataresource_1
- downloadUrl: file://tests/testfiles/input_data/material_card_al.json
  mediaType: application/vnd.dlite-parse
  configuration:
    metadata: http://www.sintef.no/calm/0.1/AluminiumMaterialCard
    driver: json
  type: ss3:AluminiumMaterialCard
  dataresource: abaqus_materialcard_al1_dataresource_2
- downloadUrl: file://tests/testfiles/input_data/material_card_concrete.json
  mediaType: application/vnd.dlite-parse
  type: ss3:ConcreteMaterialCard
  configuration:
    driver: json
    metadata: http://www.sintef.no/calm/0.1/ConcreteMaterialCard


We need to pick up the input individuals, that are the final leaves in the output from OntoFlow. 
This was done by the function get_partial_pipeline_in_spec, but this function has now been completely removed. The get_pipeline_spec was used to get this initial part in place. The generated pipeline from the previous run is attached. 
[generated_pipeline_2.txt](https://github.com/user-attachments/files/16759344/generated_pipeline_2.txt)

The correpoding file that is generated now is this: 
[pipeline_0.txt](https://github.com/user-attachments/files/16759371/pipeline_0.txt)

**Comment by louisponet** [comment]
1. The outputs here are in the correct place, this is to communicate to aiida shell which files are output files and need to be retrieved
2. This is most likely because I didn't generate these with the new KB, did you try generating them on your machine? What is the output then?

**Comment by francescalb** [comment]
> 1. The outputs here are in the correct place, this is to communicate to aiida shell which files are output files and need to be retrieved
> 2. This is most likely because I didn't generate these with the new KB, did you try generating them on your machine? What is the output then?

1. It  is wrong. It needs to be indented once.
2. I generated both with the same KB. (Current master on SS3_wrappers repo). The issue is that ontoconv no longer picks up that the files come from an individual that needs to be fetched and parsed, rather from a model excution step in ExecFlow)

File: ontoconv/ontoflow.py [hunk 2]
```diff
@@ -26,23 +149,38 @@ def parse_ontoflow(workflow_data, kb, outdir="."):
         Pipeline and workchain files are saved as yaml.
 
     """
-    outdir = Path(outdir)
-    # Fetch the pipeline specifications
-    pipelines = get_all_pipelines_spec(data=workflow_data)
```
Comment (francescalb) [inline ontoconv/ontoflow.py:31]: @louisponet 
In order to recognise when thedata comes from an already existing data set or not, a reorganisation of the pipeline was done with the get_all_pipelines_spec function. Now these are lost.

**Comment by louisponet** [comment]
1. I see, I misunderstood that you meant indent by 1 more. Fixed that.
2. Also fixed, forgot that for those guys we need to go up twice. Should be fixed too (see attached photo)! 
image

**Comment by francescalb** [comment]
Hi @louisponet , can you fix the problem with the mismatch between the output of the execwrapper and how to pick this up in the postprocessing step?

**Comment by louisponet** [comment]
Done I think

**Comment by francescalb** [comment]
> Done I think

Great, now this part works. Could you update the tests accordingly?

Then the issue with the templates and how to refer to those must be adressed in a different PR. These are crucial for Sucess stories that need templates (i.e. use Aiida wrappers)

**Comment by francescalb** [comment]
Hi @louisponet 
How is it giong with the tests? This PR should go into master soon  to make it easier to make other, needed updates.

**Comment by louisponet** [comment]
Hi Francesca,

I will be looking into this in the next days. Do you agree that it would make sense to have tests that also actually execute the chains to lead to the correct result rather than just checking verbatim the output of ontoconv? In that case maybe we can base them on the tests that exist in ExecFlow. Would it be possible to for example generate a small knowledgebase that holds the information to execute the exec_wrapper test in ExecFlow?

**Comment by louisponet** [comment]
Hi @francescalb 
For now I’ve just implemented a similar expected vs output check as before to get this merged

**Comment by francescalb** [comment]
@louisponet 

I think the test is fine for this PR, but the CI should work. To see the error messages from pre-commit you can go into the output of the gituhb actions (press the red cross when it has failed) and the error messages are quite stragiht forward as you se below:

image


To install pre-commit locally:

pip install pre-commit
pre-commit install # in the root folder where the [.pre-commit-config.yaml](https://github.com/H2020-OpenModel/OntoConv/blob/master/.pre-commit-config.yaml) is located.

# to run it locally. This is good to do as it will do some autmatic fixes for you which you then can commit.
pre-commit run --all

**Review by francescalb — APPROVED** [review]

**Comment by francescalb** [comment]
@louisponet . I have now approved it, and I think it is fine to merge this PR now.

**Pull request merged**